engine/src/main/java/org/terasology/engine/subsystem/openvr/OpenVRControllers.java

```diff
@@ -51,90 +59,83 @@
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(ControllerId.ONE),
-                    "OpenVR", ButtonState.UP, 1.0f));
-        } else if (OpenVRUtil.switchedDown(ControllerListener.BUTTON_GRIP, stateBefore.ulButtonPressed, stateAfter.ulButtonPressed)) {
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(ControllerId.ONE),
-                    "OpenVR", ButtonState.DOWN, 1.0f));
-        } else if (OpenVRUtil.switchedUp(ControllerListener.BUTTON_APP_MENU, stateBefore.ulButtonPressed, stateAfter.ulButtonPressed)) {
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(ControllerId.TWO),
-                    "OpenVR", ButtonState.UP, 1.0f));
-        } else if (OpenVRUtil.switchedDown(ControllerListener.BUTTON_APP_MENU, stateBefore.ulButtonPressed, stateAfter.ulButtonPressed)) {
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(ControllerId.TWO),
-                    "OpenVR", ButtonState.DOWN, 1.0f));
-        } else if (OpenVRUtil.switchedDown(ControllerListener.BUTTON_TOUCHPAD, stateBefore.ulButtonPressed, stateAfter.ulButtonPressed)) {
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(ControllerId.X_AXIS),
-                    "OpenVR", ButtonState.DOWN, -stateAfter.rAxis[0].x));
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(ControllerId.Y_AXIS),
-                    "OpenVR", ButtonState.DOWN, stateAfter.rAxis[0].y));
-        } else if (OpenVRUtil.switchedUp(ControllerListener.BUTTON_TOUCHPAD, stateBefore.ulButtonPressed, stateAfter.ulButtonPressed)) {
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(ControllerId.X_AXIS),
-                    "OpenVR", ButtonState.UP, 0));
-            queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(ControllerId.Y_AXIS),
-                    "OpenVR", ButtonState.UP, 0));
+    private boolean switchedUp(long buttonIndex) {
+        return OpenVRUtil.switchedUp(buttonIndex, cachedStateBefore.ulButtonPressed, cachedStateAfter.ulButtonPressed);
+    }
+
+    private boolean switchedDown(long buttonIndex) {
+        return OpenVRUtil.switchedDown(buttonIndex, cachedStateBefore.ulButtonPressed, cachedStateAfter.ulButtonPressed);
+    }
+
+    private void addAxisAction(int controllerButton, ButtonState buttonState, float axisValue) {
+        queuedActions.add(new ControllerAction(InputType.CONTROLLER_AXIS.getInput(controllerButton),
+                    "OpenVR", buttonState, axisValue));
+    }
+
+    private void addButtonAction(int controllerButton, ButtonState buttonState) {
+        queuedActions.add(new ControllerAction(InputType.CONTROLLER_BUTTON.getInput(controllerButton),
+                "OpenVR", buttonState, 1.0f));
+    }
+
+    private void handleController0() {
```

Comment:
I should clarify that the only data getting passed is a single `long` value. `isDown` and `isUp` both use `isPressed`, which is just a simple bit mask check. This allows for a long value to have the entire state of the controller. So `TRIGGER & buttonState == true` and `GRIP & buttonState == true` can simultaneously be satisfied.

The just is that even though it looks like there are a lot of function calls / work being done there the code will reduce to several bit mask checks, which is very cheap. And they all need to be checked every time.